### PR TITLE
[#106191288] Add logsearch for Cloud Foundry components and Apps

### PR DESCRIPTION
**Pull request by actionjack**
[Aggregate system logs ](https://www.pivotaltracker.com/story/show/106191288)
## What

To support the trial we need to have aggregated logs for the system to identify where issues are. We should aim for the simplest possible solution that would allow us to see the logs for all CF components and nodes in a single location to allow for easy debugging. Each log entry should identify which component and node generated the entry.

**Acceptance Criteria**
**Given** that there is an issue that I am attempting to debug
**When** look at the aggregated CF system logs
**Then** I can see all the log entries generated by all CF components
**And** each entry clearly indicates the component that generated the entry
**And** each entry clearly indicates that the node that generated the entry
## How this PR should  be reviewed

This PR should be review in the following context:
- I want to use [logsearch-for-cloudfoundry](https://github.com/logsearch/logsearch-for-cloudfoundry) logsearch addon to create a custom [Elasticsearch-Logstash-Kibana](https://www.elastic.co/products) stack deployment
- I want to use [Terraform](https://terraform.io/) to create the right network resources so I can span it across 2 Availability Zones and route all it's traffic through the bastion box.
- The current `logsearch-deployment.yml` file fails because the collector template has now moved to be part of the cf release I want to fix this.
- The current `logsearch-jobs.yml` file has lots of unnecessary white spaces that can be removed to make the code cleaner and cross IDE friendly.
- The current `logsearch-jobs.yml` deploys too many VMs, I want to have the option to deploy all the services in one VM, by adding a `logsearch-minimal-jobs.yml`
- I want to create a logsearch deployment that works for `AWS`
- I want to upload `all the things` and have a script to generate a logsearch deployment yaml file.
- I want to integrate the deployment with Makefile tasks.
- I want all the components from CF to send logs to logsearch, by setting the right configuration on the metron agents.
- I want a consistent deployment name as  `logsearch-<envname>`
- I want to create a logsearch deployment that works for `GCE`, for which I need to use DNS to point the services instead of static IPs.
- I want to be able to deploy `kibana-for-cloudfoundry` using the errand from `logsearch-for-cloudfoundry`, for what I need set the IP of elasticsearch server in the manifest after deploying it.
- I want `collector` to connect to nats on `AWS` and `GCE`, using DNS on GCE instead of static IPs.
## How to Test this PR
- Deploy an environment in both AWS and GCE (it is OK if you already have an environment):
  - AWS `make aws DEPLOY_ENV=$MYENVIRONMENT`
  - GCE `make aws DEPLOY_ENV=$MYENVIRONMENT`
- Connect using ssh to the bastion box (`make ssh-aws` or `make ssh-gce`)
- To deploy, run the following on the bastion after deploying CF:

```
./scripts/deploy_logsearch.sh aws
```

  or 

```
./scripts/deploy_logsearch.sh gce
```
- Access the board in: https://logs.$MYENVIRONMENT.cf.paas.alphagov.co.uk/ or https://logs.$MYENVIRONMENT.cf2.paas.alphagov.co.uk/ 
## Who should review this PR
- A handsome hedgehog, if a good looking spiny mammal is unavailable anyone in the core team can step up to the plate except @keymon or @actionjack

### CODE REVIEW

File: manifests/templates/logsearch/logsearch-deployment.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    version: latest
+  - name: logsearch-for-cloudfoundry
+    version: latest
+  - name: cf
+    version: latest
+  default_env:
+    # Default vcap & root password on deployed VMs (ie c1oudc0w)
+    # Generated using mkpasswd -m sha-512
+    bosh:
+      password: (( merge || "$6$4gDD3aV0rdqlrKC$2axHCxGKIObs6tAmMTqYCspcdvQXh3JJcvWOY2WGb4SrdXtnCyNaWlrf3WEqvYR2MYizEGp3kMmbpwBC6jsHt0" ))
```
Comment (dcarley) [inline manifests/templates/logsearch/logsearch-deployment.yml:16]: Should this be in pass? Are we overriding it? If so, should we omit the default?

File: manifests/templates/logsearch/logsearch-infrastructure-warden.yml [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    name: random
+- name: high_memory
+  cloud_properties:
+    name: random
+- name: high_cpu
+  cloud_properties:
+    name: random
+
+networks: (( merge || networks_default ))
+networks_default:
```
Comment (dcarley) [inline manifests/templates/logsearch/logsearch-infrastructure-warden.yml:45]: I'm confused; what are all of these networks listed below?

File: manifests/templates/logsearch/logsearch-deployment.yml (hunk 1 above)
Comment (keymon) [inline manifests/templates/logsearch/logsearch-deployment.yml:16]: Good catch, actually we missed overriding the passwords properly. I will fix that.

File: manifests/templates/logsearch/logsearch-infrastructure-warden.yml (hunk 2 above)
Comment (keymon) [inline manifests/templates/logsearch/logsearch-infrastructure-warden.yml:45]: This is indeed really confusing. This is actually a workaround of a issue in the warden bosh CPI, which forces you to define all those networks.

This is only required for warden, and we don't really use it, unless you deploy this on bosh-lite.

File: manifests/templates/logsearch/logsearch-deployment.yml (hunk 1 above)
Comment (keymon) [inline manifests/templates/logsearch/logsearch-deployment.yml:16]: Fixed, I override that. 

Do you want to omit the default?

File: manifests/templates/logsearch/logsearch-infrastructure-aws.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+- name: high_memory
+  cloud_properties:
+    instance_type: r3.large
+    ephemeral_disk:
+      size: 10_240
+      type: gp2
+    availability_zone: (( meta.zones.z1 ))
+- name: high_cpu
+  cloud_properties:
+    instance_type: r3.large
```
Comment (mtekel) [inline manifests/templates/logsearch/logsearch-infrastructure-aws.yml]: How about c4.large? r3 instances are high memory, r3.large has 15G ram. C4 are compute, better corresponding to the GCE highcpu instances...

Comment (keymon) [inline manifests/templates/logsearch/logsearch-infrastructure-aws.yml]: It makes sense I will change it. Right we only use the high_memory pool, as we have only one instance using the `logsearch-minimal-jobs.yml` template.

Comment (keymon) [inline manifests/templates/logsearch/logsearch-infrastructure-aws.yml]: done

**Comment by mtekel** [comment]
There's one small nuisance this creates: When you have 2 deployments, bosh will have the last deployment selected - which is logsearch. You will not be able to do BOSH cli operations on the jobs from the main deployment until you select it. For that you need a deployment file. Luckily if you are on bastion the file is there (cf-manifest.yml). Otherwise you can still download manifest using bosh.

**Comment by mtekel** [comment]
Works on GCE and AWS as well, merging.

**Pull request merged**